### PR TITLE
FLUME-3050 add counters for error conditions and expose to monitor URL

### PR DESCRIPTION
**Pull request by majorendre**
By introducing error counters it will be easier to monitor problems. Also errors are categorized, hopefully this will help setting up better monitoring solutions.
Concept: an error is when an Exception is thrown or an ERROR level log is written during event processing.
In case of an error at least 1 error counter is increased at least once. (Preferably 1 counter once).
Errors during event processing are counted. Initialization errors are not handled here.
3 types of errors are differentiated.
-Channel read/write errors from the channel when the channel throws a ChannelException.
-Event read/write errors. E.g: A source cannot read an event due to 
-Generic errors - e.g.: TaildirSource cannot write position file.

### CODE REVIEW

**Review by turcsanyip — APPROVED** [review]
It looks good to me. The tests passed.
Thanks for the contribution.

File: flume-ng-core/src/test/java/org/apache/flume/source/TestAvroSource.java [hunk 1]
```diff
@@ -64,9 +67,16 @@
...
 import org.junit.Test;
+import org.mockito.Mockito;
+import org.mockito.internal.util.reflection.Whitebox;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyListOf;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.when;
```
Comment (szaboferee) [inline flume-ng-core/src/test/java/org/apache/flume/source/TestAvroSource.java]: unused import

**Review by szaboferee — CHANGES_REQUESTED** [review]
Thanks for the contribution @majorendre 

nice change, well tested. 
Could you add the counters to the SequenceGeneratorSource, HiveSink and KafkaSink?
Also it would be great to have a test for the SpoolDirSource as it looks like that it may count some errors both in the Source and the ReliableSpoolingFileEventReader.

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by majorendre** [comment]
Rewrote the SpoolDirectorySource and ReliableSpoolingFileEventReader part.
Only increment the counters in ReliableSpoolingFileEventReader when the problem is handled with logging an error message only. When an exception is thrown, then it will be handled in SpoolDirectorySource, it is much clearer this way. This was one of the first ones I did, and the concept was not ready, yet. Created some Mockito test cases for SpoolDirectorySource changes but I did not create for the ReliableSpoolingFileEventReader part, since it is so simple, and would be quite difficult to test.
I did all the other requested changes, too.

**Comment by majorendre** [comment]
retest this please

**Pull request closed without merging**